mlir/lib/Target/IRDLToCpp/IRDLToCpp.cpp

```diff
@@ -0,0 +1,530 @@
+    return failure();
+  const auto forwardDeclarations = llvm::formatv(
+      R"(
+{1}
+{0}
+{2}
+    )",
+      llvm::join(llvm::map_range(opNames,
+                                 [](llvm::StringRef name) -> std::string {
+                                   return llvm::formatv("class {0};", name);
+                                 }),
+                 "\n"),
+      dialectStrings.namespaceOpen, dialectStrings.namespaceClose);
+
+  output << forwardDeclarations;
+  for (auto &&operationOp : operationOps) {
+    if (failed(generateOperationInclude(operationOp, output, dict)))
+      return failure();
+  }
+
+  output << "#endif // " << declarationMacroFlag << "\n";
+
+  return success();
+}
+
+static LogicalResult generateLib(irdl::DialectOp dialect, raw_ostream &output,
+                                 DialectStrings &dialectStrings) {
+
+  static const auto typeHeaderDefTemplate = mlir::irdl::detail::Template{
+#include "Templates/TypeHeaderDef.txt"
+  };
+  static const auto opDefTemplate = mlir::irdl::detail::Template{
+#include "Templates/OperationDef.txt"
+  };
+  static const auto typeDefTemplate = mlir::irdl::detail::Template{
+#include "Templates/TypeDef.txt"
+  };
+  static const auto dialectDefTemplate = mlir::irdl::detail::Template{
+#include "Templates/DialectDef.txt"
+  };
+  static const auto perOpDefTemplate = mlir::irdl::detail::Template{
+#include "Templates/PerOperationDef.txt"
+  };
+
+  irdl::detail::dictionary dict;
+  fillDict(dict, dialectStrings);
+
+  output << "#ifdef " << definitionMacroFlag << "\n#undef "
+         << definitionMacroFlag << "\n";
+
+  typeHeaderDefTemplate.render(output, dict);
+
+  // get typedef list
+  llvm::SmallVector<std::string> typeNames;
+  if (failed(generateTypedefList(dialect, typeNames)))
+    return failure();
+
+  dict["TYPE_LIST"] = llvm::join(
+      llvm::map_range(typeNames,
+                      [&dialectStrings](llvm::StringRef name) -> std::string {
+                        return llvm::formatv(
+                            "{0}::{1}", dialectStrings.namespacePath, name);
+                      }),
+      ",\n");
+
+  dict["TYPE_PARSER"] = llvm::formatv(
+      R"(static ::mlir::OptionalParseResult generatedTypeParser(::mlir::AsmParser &parser, ::llvm::StringRef *mnemonic, ::mlir::Type &value) {
+  return ::mlir::AsmParser::KeywordSwitch<::mlir::OptionalParseResult>(parser)
+    {0}    
+    .Default([&](llvm::StringRef keyword, llvm::SMLoc) {{
+      *mnemonic = keyword;
+      return std::nullopt;
+    });
+})",
+      llvm::join(
+          llvm::map_range(
+              typeNames,
+              [&](llvm::StringRef name) -> std::string {
+                return llvm::formatv(
+                    R"(.Case({1}::{0}::getMnemonic(), [&](llvm::StringRef, llvm::SMLoc) {
+      value = {1}::{0}::get(parser.getContext());
+      return ::mlir::success(!!value);
+    }))",
+                    name, dialectStrings.namespacePath);
+              }),
+          "\n"));
```

Comment:
The nesting is getting a bit crazy, could you break this down into local variables? :) Feel free to do it in other places where it's less dramatic too.